samples/WinFormsDx12/Dx12Viewport.cs

```diff
@@ -0,0 +1,684 @@
+                    {
+                        var hr = Marshal.GetHRForLastWin32Error();
+                        Marshal.ThrowExceptionForHR(hr);
+                    }
+
+                    // Wait for the command list to execute; we are reusing the same command
+                    // list in our main loop but for now, we just want to wait for setup to
+                    // complete before continuing.
+                    WaitForPreviousFrame();
+                }
+            }
+            finally
+            {
+                if (signature != null)
+                {
+                    signature->Release();
+                }
+
+                if (error != null)
+                {
+                    error->Release();
+                }
+
+                if (vertexShader != null)
+                {
+                    vertexShader->Release();
+                }
+
+                if (pixelShader != null)
+                {
+                    pixelShader->Release();
+                }
+            }
+        }
+
+        private void PopulateCommandList()
+        {
+            // Command list allocators can only be reset when the associated
+            // command lists have finished execution on the GPU; apps should use
+            // fences to determine GPU execution progress.
+            ThrowIfFailed(nameof(ID3D12CommandAllocator.Reset), _commandAllocator->Reset());
+
+            // However, when ExecuteCommandList() is called on a particular command
+            // list, that command list can then be reset at any time and must be before
+            // re-recording.
+            ThrowIfFailed(nameof(ID3D12GraphicsCommandList.Reset), _commandList->Reset(_commandAllocator, _pipelineState));
+
+            // Set necessary state.
+            _commandList->SetGraphicsRootSignature(_rootSignature);
+
+            fixed (D3D12_VIEWPORT* viewport = &_viewport)
+            {
+                _commandList->RSSetViewports(1, viewport);
+            }
+
+            fixed (RECT* scissorRect = &_scissorRect)
+            {
+                _commandList->RSSetScissorRects(1, scissorRect);
+            }
+
+            // Indicate that the back buffer will be used as a render target.
+            var barrier = D3D12_RESOURCE_BARRIER.InitTransition(_renderTargets[unchecked((int)_frameIndex)], D3D12_RESOURCE_STATE_PRESENT, D3D12_RESOURCE_STATE_RENDER_TARGET);
+            _commandList->ResourceBarrier(1, &barrier);
+
+            var rtvHandle = _rtvHeap->GetCPUDescriptorHandleForHeapStart();
+            rtvHandle.ptr = (UIntPtr)((byte*)rtvHandle.ptr + (_frameIndex * _rtvDescriptorSize));
+            _commandList->OMSetRenderTargets(1, &rtvHandle, FALSE, null);
+
+            // Record commands.
+            var r = _random.Next(0, 50) / 100.0f;
+            var g = _random.Next(0, 50) / 100.0f;
+            var b = _random.Next(0, 50) / 100.0f;
+            var clearColor = stackalloc float[4] { 0.0f, r, g, b };
```

Comment:
It would be better to either toggle between two well-defined colors or use a color picker control and letting the user select the target color

Reply:
Implemented toggle between two colors. The idea is just to be able to see that it indeed re-rendered the scene in response to the button or resize event.